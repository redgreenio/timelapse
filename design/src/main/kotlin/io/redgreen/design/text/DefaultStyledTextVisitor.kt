package io.redgreen.design.text

open class DefaultStyledTextVisitor : StyledTextVisitor {
  override fun onEnterLine(lineNumber: Int) {
    /* override if necessary :) */
  }

  override fun onExitLine(lineNumber: Int) {
    /* override if necessary :) */
  }

  override fun onEnterLine(lineNumber: Int, style: LineStyle) {
    /* override if necessary :) */
  }
}
