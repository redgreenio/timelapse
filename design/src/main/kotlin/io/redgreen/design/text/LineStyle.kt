package io.redgreen.design.text

data class LineStyle(
  val name: String,
  val lineNumberRange: IntRange
) {
  constructor(name: String, lineNumber: Int) : this(name, lineNumber..lineNumber)
}
