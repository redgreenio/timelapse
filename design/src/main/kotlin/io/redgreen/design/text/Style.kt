package io.redgreen.design.text

data class LineStyle(
  val name: String,
  val lineNumberRange: IntRange
) {
  constructor(name: String, lineNumber: Int) : this(name, lineNumber..lineNumber)
}

data class TextStyle(
  val name: String,
  val lineNumber: Int,
  val charIndexRange: IntRange
) {
  constructor(name: String, lineNumber: Int, charIndex: Int) : this(name, lineNumber, charIndex..charIndex)
}
