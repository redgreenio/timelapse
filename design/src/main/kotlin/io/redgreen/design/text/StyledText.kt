package io.redgreen.design.text

import java.lang.StringBuilder
import java.util.Optional

data class StyledText(val text: String) {
  private val lineStyles = mutableListOf<LineStyle>()

  fun visit(visitor: StyledTextVisitor) {
    var lineNumber = 1

    getLineStyle(lineNumber)
      .ifPresentOrElse(
        { visitor.onEnterLine(lineNumber, it) },
        { visitor.onEnterLine(lineNumber) }
      )

    val textBuilder = StringBuilder()

    text.onEachIndexed { index, char ->
      if (char == '\n') {
        lineNumber++

        getLineStyle(lineNumber)
          .ifPresentOrElse(
            { visitor.onEnterLine(lineNumber, it) },
            { visitor.onEnterLine(lineNumber) }
          )
      } else {
        if (char != '\n') {
          textBuilder.append(char)
        }
      }

      if ((index + 1 > text.lastIndex) || (index + 1 != text.lastIndex && text[index + 1] == '\n')) {
        visitor.onText(textBuilder.toString())
        textBuilder.clear()

        getLineStyle(lineNumber)
          .ifPresentOrElse(
            { visitor.onExitLine(lineNumber, it) },
            { visitor.onExitLine(lineNumber) }
          )
      }
    }
  }

  private fun getLineStyle(lineNumber: Int): Optional<LineStyle> {
    return Optional.ofNullable(lineStyles.find { lineNumber in it.lineNumberRange })
  }

  fun addStyle(lineStyle: LineStyle): StyledText {
    lineStyles.add(lineStyle)
    return this
  }
}
