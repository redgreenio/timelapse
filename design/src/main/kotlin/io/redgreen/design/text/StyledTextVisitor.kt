package io.redgreen.design.text

interface StyledTextVisitor {
  fun onEnterLine(lineNumber: Int)
  fun onExitLine(lineNumber: Int)
  fun onEnterLine(lineNumber: Int, style: LineStyle)
}
