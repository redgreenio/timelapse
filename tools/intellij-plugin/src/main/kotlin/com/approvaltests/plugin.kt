package com.approvaltests

const val ID = "io.redgreen.timelapse.dev"
const val DISPLAY_NAME = "Approval Tests"
