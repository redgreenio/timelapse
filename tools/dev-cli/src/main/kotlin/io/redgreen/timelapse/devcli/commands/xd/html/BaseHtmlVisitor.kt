package io.redgreen.timelapse.devcli.commands.xd.html

import io.redgreen.design.text.LineStyle
import io.redgreen.design.text.StyledTextVisitor
import io.redgreen.design.text.TextStyle
import kotlin.LazyThreadSafetyMode.NONE

class BaseHtmlVisitor : StyledTextVisitor {
  companion object {
    private const val INDENT = "  "
    private const val NEWLINE = "\n"

    private const val MARKER_TABLE_ROWS = "{table-rows}"
  }

  private val contentBuilder = StringBuilder()

  private val template by lazy(NONE) {
    BaseHtmlVisitor::class.java.classLoader
      .getResourceAsStream("template.html")!!
      .reader()
      .readText()
  }

  val content: String
    get() = template.replace(MARKER_TABLE_ROWS, contentBuilder.toString())

  override fun onText(text: String) {
    contentBuilder.append(text)
  }

  override fun onEnterLine(lineNumber: Int) {
    if (lineNumber != 1) {
      contentBuilder.append(NEWLINE)
    }
    contentBuilder
      .append("<tr>")
      .append(NEWLINE)
      .append("""$INDENT<td class="line-number">$lineNumber</td>""")
      .append(NEWLINE)
      .append("$INDENT<td>")
  }

  override fun onExitLine(lineNumber: Int) {
    contentBuilder
      .append("</td>")
      .append(NEWLINE)
      .append("</tr>")
  }

  override fun onEnterLine(lineNumber: Int, lineStyle: LineStyle) {
    // no-p
  }

  override fun onExitLine(lineNumber: Int, lineStyle: LineStyle) {
    // no-p
  }

  override fun onBeginStyle(textStyle: TextStyle) {
    // no-p
  }

  override fun onEndStyle(textStyle: TextStyle) {
    // no-p
  }
}
