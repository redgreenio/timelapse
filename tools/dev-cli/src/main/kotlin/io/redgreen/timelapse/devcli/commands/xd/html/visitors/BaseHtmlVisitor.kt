package io.redgreen.timelapse.devcli.commands.xd.html.visitors

import io.redgreen.design.text.LineStyle
import io.redgreen.design.text.StyledTextVisitor
import io.redgreen.design.text.TextStyle
import io.redgreen.timelapse.devcli.commands.TOOL_VERSION
import org.apache.commons.text.StringEscapeUtils
import kotlin.LazyThreadSafetyMode.NONE

class BaseHtmlVisitor(
  private val title: String,
  private val affectedLineNumbers: List<Int> = emptyList()
) : StyledTextVisitor {
  companion object {
    private const val INDENT = "  "
    private const val NEWLINE = "\n"
    private const val CHAR_SPACE = ' '
    private const val CHAR_TAB = '\t'

    private const val NBSP = "&nbsp;"
    private const val TAB = "$NBSP$NBSP$NBSP$NBSP"

    private const val MARKER_TITLE = "{title}"
    private const val MARKER_TABLE_ROWS = "{table-rows}"
    private const val MARKER_VERSION = "{tool-version}"
  }

  private val contentBuilder = StringBuilder()

  private val template by lazy(NONE) {
    BaseHtmlVisitor::class.java.classLoader
      .getResourceAsStream("template.html")!!
      .reader()
      .readText()
  }

  val content: String
    get() {
      val rawTableRows = contentBuilder.toString()
      val indentedTableRows = rawTableRows.split(NEWLINE).joinToString(NEWLINE) { "$INDENT$it" }

      return template
        .replace(MARKER_TITLE, title)
        .replace(MARKER_TABLE_ROWS, indentedTableRows)
        .replace(MARKER_VERSION, TOOL_VERSION)
    }

  override fun onText(text: String) {
    contentBuilder.append(toHtmlFriendly(text))
  }

  override fun onEnterLine(lineNumber: Int) {
    if (lineNumber != 1) {
      contentBuilder.append(NEWLINE)
    }
    contentBuilder
      .append("<tr>")
      .append(NEWLINE)
      .append("""$INDENT<td class="line-number">$lineNumber</td>""")
      .append(NEWLINE)

    val muted = lineNumber !in affectedLineNumbers
    if (muted) {
      contentBuilder.append("""$INDENT<td class="muted">""")
    } else {
      contentBuilder.append("""$INDENT<td>""")
      if (isInsideMultilineString) {
        contentBuilder.append("<span class=\"string\">")
      }
    }
  }

  override fun onExitLine(lineNumber: Int) {
    if (isInsideMultilineString) {
      contentBuilder.append("</span>")
    }

    contentBuilder
      .append("</td>")
      .append(NEWLINE)
      .append("</tr>")
  }

  override fun onEnterLine(lineNumber: Int, lineStyle: LineStyle) {
    if (isInsideMultilineString) {
      contentBuilder.append("<span class=\"string\">")
    }
  }

  override fun onExitLine(lineNumber: Int, lineStyle: LineStyle) {
    if (isInsideMultilineString) {
      contentBuilder.append("</span>")
    }
  }

  private var isInsideMultilineString = false

  override fun onBeginStyle(textStyle: TextStyle) {
    if (textStyle.name == "open-multiline-string") {
      isInsideMultilineString = true
    }

    if (textStyle.name == "close-multiline-string") {
      isInsideMultilineString = false
    }

    if (textStyle.name != "end-string") {
      contentBuilder.append("""<span class="${textStyle.name}">""")
    }
  }

  override fun onEndStyle(textStyle: TextStyle) {
    if (textStyle.name != "string") {
      contentBuilder.append("</span>")
    }

    if (textStyle.name == "open-multiline-string" && isInsideMultilineString) {
      contentBuilder.append("<span class=\"string\">")
    } else if (textStyle.name == "close-multiline-string") {
      contentBuilder.append("</span>")
    }
  }

  private fun toHtmlFriendly(line: String): String {
    val startSpaceCharsTrimmedLine = line.trimStart(CHAR_SPACE)
    val nStartSpaceChars = line.length - startSpaceCharsTrimmedLine.length

    val htmlEscapedLine = StringEscapeUtils.escapeHtml4(line).trimStart(CHAR_SPACE)
    return (1..nStartSpaceChars)
      .joinToString("") { NBSP } + htmlEscapedLine
      .replace("$CHAR_TAB", TAB)
  }
}
