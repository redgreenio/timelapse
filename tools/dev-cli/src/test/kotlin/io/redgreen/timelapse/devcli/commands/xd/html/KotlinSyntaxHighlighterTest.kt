package io.redgreen.timelapse.devcli.commands.xd.html

import io.redgreen.design.text.StyledText
import io.redgreen.timelapse.devcli.commands.xd.html.visitors.BaseHtmlVisitor
import org.approvaltests.Approvals
import org.junit.jupiter.api.Test

class KotlinSyntaxHighlighterTest {
  @Test
  fun `it should highlight Kotlin source code`() {
    // given
    val sourceCode = """
      package io.redgreen.timelapse.devcli.commands.xd.html

      sealed class GitCommand {
        object LsFiles : GitCommand() {
          fun command(fileName: String): ShellCommand =
            ShellCommand(arrayOf("git", "ls-files", fileName, "**/${'$'}fileName"))
        }

        object Show : GitCommand() {
          fun command(commitHash: String, filePath: String): ShellCommand =
            ShellCommand(arrayOf("git", "show", "${'$'}commitHash:${'$'}filePath"))
        }

        object Status : GitCommand() {
          fun command(): ShellCommand =
            ShellCommand(arrayOf("git", "status"))
        }

        object GetUnifiedPatch : GitCommand() {
          fun command(commitHash: String, filePath: String): ShellCommand =
            ShellCommand(arrayOf("git", "diff", "-u", "${'$'}commitHash^1", commitHash, "--", filePath))
        }
      }
    """.trimIndent()

    val styledText = StyledText(sourceCode)

    val affectedLineNumbers = (1..23).toList()

    // when
    KotlinSyntaxHighlighter.addStylesForTokens(styledText, affectedLineNumbers)
    val visitor = BaseHtmlVisitor("Kotlin Source", affectedLineNumbers)
    styledText.visit(visitor)

    // then
    Approvals.verifyHtml(visitor.content)
  }
}
