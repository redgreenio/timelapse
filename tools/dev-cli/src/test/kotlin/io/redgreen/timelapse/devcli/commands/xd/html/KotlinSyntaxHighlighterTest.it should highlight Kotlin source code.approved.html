<!DOCTYPE html>
<!--suppress CssUnusedSymbol -->
<!--Generated by dev-cli: v0.0.14-dev-->
<html lang="en-US">

<head>
  <style>
    @import url('https://fonts.googleapis.com/css2?family=JetBrains+Mono&display=swap');
  </style>
  <script src="https://code.jquery.com/jquery-2.2.4.min.js"
          integrity="sha256-BbhdlvQf/xTY9gja0Dq3HiwQF8LaCRTXxZKRutelT44=" crossorigin="anonymous"></script>
  <style>
    body {
      font-family: 'JetBrains Mono', monospace;
      font-size: small;
      margin-left: 0;
      margin-right: 0;
    }

    table {
      border-collapse: collapse;
    }

    tr td {
      padding-left: 0.5em;
    }

    .line-number {
      color: #9F9F9F;
      min-width: 3em;
    }

    .deleted-line {
      background-color: #FFDCE0;
    }

    .deleted-line .no-line-number {
      background-color: #FFB4BC;
    }

    .affected-line {
      font-weight: normal;
    }

    .keyword {
      color: #000080;
    }

    .comment {
      color: #6D6D6D;
    }

    td:nth-child(2) {
      padding: 0.18em 0.0em 0.18em 0.8em;
      width: 100%;
    }

    .parentheses {
      color: #19561E;
      font-weight: normal;
    }

    .curly {
      color: #0058BE;
      font-weight: normal;
    }

    .modified-line {
      background-color: #F1F8FF;
    }

    .modified-line .line-number {
      background-color: #DBEDFF;
    }

    .modified-highlight {
      background-color: #3B9DFF4D;
      margin: 0;
      padding: 0 0.18em 0 0.18em;
      border-radius: 0.2em;
    }

    .added-line {
      background-color: #E6FFED;
    }

    .added-line .line-number {
      background-color: #C5FFD5;
    }

    .added-highlight {
      background-color: #C5FFD5;
      margin: 0;
      padding: 0 0.18em 0 0.18em;
      border-radius: 0.2em;
    }

    .string {
      color: #00733B;
    }

    .readability-line {
      background-color: #F8F8F8;
    }

    .readability-line .no-line-number {
      background-color: #EEEEEE;
    }

    .predecessor {
      font-weight: normal;
      color: #6D6D6D;
      text-decoration: line-through;
    }

    .change-count {
      color: #3F3F3F;
      background-color: #FFD54F;
      text-align: center;
      padding: 0.0em 0.25em 0.0em 0.25em;
      border-radius: 0.3em;
      font-size: 0.9em;
      font-weight: normal;
    }

    .annotation {
      color: #808000;
    }

    .function-signature ~ .function-body {
      display: none;
    }

    .open .function-signature ~ .function-body {
      display: table-row;
    }

    .function-signature .identifier {
      cursor: pointer;
    }

    .identifier {
      color: #000000;
    }

    .function-signature .collapsed-body-indicator {
      display: inline;
    }

    .open .function-signature .collapsed-body-indicator {
      display: none;
    }

    .ellipsis {
      padding-left: 0.2em;
      padding-right: 0.2em;
    }

    .muted {
      color: #A0A0A0;
    }

    .property {
      color: #660E7A;
    }

    .const {
      color: #660E7A;
    }

    .number {
      color: #007EE9;
    }

    .escape-sequence {
      color: #000080;
    }

    .reserved-symbol {
      color: #000080;
    }

    .deleted-highlight {
      background-color: #FFB4BC;
      margin: 0;
      padding: 0 0.18em 0 0.18em;
      border-radius: 0.2em;
    }

    .xd-comment {
      background-color: rgba(200, 200, 200, 0.4);
      color: rgba(0, 0, 0, 0.65);
      padding: 0 0.4em 0 0.2em;
      border-radius: 0.18em;
    }

    .xd-comment::before {
      content: "ef";
      font-size: 0.8em;
      color: white;
      margin-right: 0.4em;
      background-color: red;
      font-weight: bold;
      padding: 0.15em 0.4em 0.15em 0.4em;
      border-radius: 0.18em 0 0 0.18em;
      position: relative;
      top: -0.1em;
      left: -0.25em;
    }

    .todo {
      color: #1818E9;
      font-style: italic;
    }

    .todo .parentheses {
      color: #005043;
      font-style: italic;
    }
  </style>
  <script>
    $(document).ready(function () {
      $('table').on('click', 'tr.function-signature .identifier', function () {
        $(this).closest('tbody').toggleClass('open');
      })
    })
  </script>
  <title>Kotlin Source</title>
</head>

<body>
<table>
  <tbody>
  <tr>
    <td class="line-number">1</td>
    <td><span class="keyword">package</span>&nbsp;io.redgreen.timelapse.devcli.commands.xd.html</td>
  </tr>
  <tr>
    <td class="line-number">2</td>
    <td></td>
  </tr>
  <tr>
    <td class="line-number">3</td>
    <td><span class="keyword">sealed</span>&nbsp;<span class="keyword">class</span>&nbsp;GitCommand {</td>
  </tr>
  <tr>
    <td class="line-number">4</td>
    <td>&nbsp;&nbsp;<span class="keyword">object</span>&nbsp;LsFiles : GitCommand() {</td>
  </tr>
  <tr>
    <td class="line-number">5</td>
    <td>&nbsp;&nbsp;&nbsp;&nbsp;<span class="keyword">fun</span>&nbsp;command(fileName: String): ShellCommand =</td>
  </tr>
  <tr>
    <td class="line-number">6</td>
    <td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ShellCommand(arrayOf(&quot;git&quot;, &quot;ls-files&quot;, fileName, &quot;**/$fileName&quot;))</td>
  </tr>
  <tr>
    <td class="line-number">7</td>
    <td>&nbsp;&nbsp;}</td>
  </tr>
  <tr>
    <td class="line-number">8</td>
    <td></td>
  </tr>
  <tr>
    <td class="line-number">9</td>
    <td>&nbsp;&nbsp;<span class="keyword">object</span>&nbsp;Show : GitCommand() {</td>
  </tr>
  <tr>
    <td class="line-number">10</td>
    <td>&nbsp;&nbsp;&nbsp;&nbsp;<span class="keyword">fun</span>&nbsp;command(commitHash: String, filePath: String): ShellCommand =</td>
  </tr>
  <tr>
    <td class="line-number">11</td>
    <td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ShellCommand(arrayOf(&quot;git&quot;, &quot;show&quot;, &quot;$commitHash:$filePath&quot;))</td>
  </tr>
  <tr>
    <td class="line-number">12</td>
    <td>&nbsp;&nbsp;}</td>
  </tr>
  <tr>
    <td class="line-number">13</td>
    <td></td>
  </tr>
  <tr>
    <td class="line-number">14</td>
    <td>&nbsp;&nbsp;<span class="keyword">object</span>&nbsp;Status : GitCommand() {</td>
  </tr>
  <tr>
    <td class="line-number">15</td>
    <td>&nbsp;&nbsp;&nbsp;&nbsp;<span class="keyword">fun</span>&nbsp;command(): ShellCommand =</td>
  </tr>
  <tr>
    <td class="line-number">16</td>
    <td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ShellCommand(arrayOf(&quot;git&quot;, &quot;status&quot;))</td>
  </tr>
  <tr>
    <td class="line-number">17</td>
    <td>&nbsp;&nbsp;}</td>
  </tr>
  <tr>
    <td class="line-number">18</td>
    <td></td>
  </tr>
  <tr>
    <td class="line-number">19</td>
    <td>&nbsp;&nbsp;<span class="keyword">object</span>&nbsp;GetUnifiedPatch : GitCommand() {</td>
  </tr>
  <tr>
    <td class="line-number">20</td>
    <td>&nbsp;&nbsp;&nbsp;&nbsp;<span class="keyword">fun</span>&nbsp;command(commitHash: String, filePath: String): ShellCommand =</td>
  </tr>
  <tr>
    <td class="line-number">21</td>
    <td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ShellCommand(arrayOf(&quot;git&quot;, &quot;diff&quot;, &quot;-u&quot;, &quot;$commitHash^1&quot;, commitHash, &quot;--&quot;, filePath))</td>
  </tr>
  <tr>
    <td class="line-number">22</td>
    <td>&nbsp;&nbsp;}</td>
  </tr>
  <tr>
    <td class="line-number">23</td>
    <td>}</td>
  </tr>
  </tbody>
</table>
</body>
</html>
